drivers/regulator/regulator_pmic.c

```diff
@@ -236,6 +278,51 @@ int regulator_get_current_limit(const struct device *dev)
 	return data->current_levels[i].uA;
 }
 
+/**
+ * Part of the extended regulator consumer API
+ * switches the regulator to a given mode
+ */
+int regulator_set_mode(const struct device *dev, uint32_t mode)
+{
+	const struct regulator_config *config = dev->config;
+	struct regulator_data *data = dev->data;
+	int rc;
+	uint8_t i, sel_off;
+
+	if (config->num_modes == 0) {
+		return -ENOTSUP;
+	}
+
+	/* Search for mode ID in allowed modes. */
+	for (i = 0 ; i < config->num_modes; i++) {
+		if (config->allowed_modes[i] == mode) {
+			break;
+		}
+	}
+	if (i == config->num_modes) {
+		/* Mode was not found */
+		return -EINVAL;
+	}
+	sel_off = ((mode & PMIC_I2C_MODE_IDX_MASK) >> PMIC_I2C_MODE_IDX_SHIFT);
+	/* Configure mode */
+	if (mode & PMIC_I2C_OFFSET_MODE) {
+		/* Select mode with offset calculation */
+		data->reg_offset = sel_off;
+		rc = regulator_modify_register(dev, config->modesel_reg,
+			mode & PMIC_I2C_MODE_MASK, config->modesel_mask);
+	} else {
+		/* Select mode via I2C */
+		/* Clear register offset */
+		data->reg_offset = 0;
+		rc = regulator_modify_register(dev, config->modesel_reg,
+			mode & PMIC_I2C_MODE_MASK, config->modesel_mask);
+		if (rc) {
+			return rc;
+		}
+		data->reg_offset = sel_off;
```

Comment:
Why is this line needed?

Reply:
For I2C mode selection (without offset being applied to `modesel_reg`), we first clear the register offset, and make the mode selection. Then, the mode selection offset is applied. For offset based selection (pin mode on the PCA9420), we apply the offset first, since we want it to be applied to `modesel_reg`